src/xml_parser.h

```diff
@@ -317,6 +317,43 @@ cr_xml_parse_main_metadata_together(const char *primary_path,
                                     gboolean allow_out_of_order,
                                     GError **err);
 
+typedef struct {
+    CR_FILE *primary_f;
+    CR_FILE *filelists_f;
+    CR_FILE *other_f;
+
+    const char *primary_path;
+    const char *filelists_path;
+    const char *other_path;
+
+    void *primary_pd;
+    void *filelists_pd;
+    void *other_pd;
+
+    gboolean allow_out_of_order;
+    GError *tmp_err;
+
+    gboolean primary_is_done;
+    gboolean filelists_is_done;
+    gboolean other_is_done;
+
+    void *cbdata;
+} cr_PkgIterator;
```

Comment:
I think it should be possible to make the content private.

Here it would be just:  `typedef struct _cr_PkgIterator cr_PkgIterator;`

And at the top xml_parser_main_metadata_together.c:
```c
struct _cr_PkgIterator {
    CR_FILE *primary_f;
    CR_FILE *filelists_f;
    CR_FILE *other_f;

    const char *primary_path;
    const char *filelists_path;
    const char *other_path;

    void *primary_pd;
    void *filelists_pd;
    void *other_pd;

    gboolean allow_out_of_order;
    GError *tmp_err;

    gboolean primary_is_done;
    gboolean filelists_is_done;
    gboolean other_is_done;

    void *cbdata;
};
```
(no `typedef` here)

AFAIK this should work, it is the same as for example `_cr_ChecksumCtx`.

Reply:
I actually tried this already but I get a bunch of errors like this:

```
  /home/vagrant/devel/community/createrepo_c/src/xml_parser_main_metadata_together.c:530:13: error: invalid use of incomplete typedef ‘cr_PkgIterator’ {aka ‘struct _cr_PkgIterator’}                                                       
      530 |     if (iter->tmp_err) {                                                                                                                                                                                                          
          |             ^~                                                                                                                                                                                                                    
    /home/vagrant/devel/community/createrepo_c/src/xml_parser_main_metadata_together.c:533:22: error: invalid use of incomplete typedef ‘cr_PkgIterator’ {aka ‘struct _cr_PkgIterator’}                                                       
      533 |         cr_close(iter->primary_f, NULL);                                                                                                                                                                                          
          |                      ^~                                                                                                                                                                                                           
    /home/vagrant/devel/community/createrepo_c/src/xml_parser_main_metadata_together.c:534:22: error: invalid use of incomplete typedef ‘cr_PkgIterator’ {aka ‘struct _cr_PkgIterator’}                                                       
      534 |         cr_close(iter->filelists_f, NULL);                                                                                                                                                                                        
          |                      ^~                                                                                                                                                                                                           
    /home/vagrant/devel/community/createrepo_c/src/xml_parser_main_metadata_together.c:535:22: error: invalid use of incomplete typedef ‘cr_PkgIterator’ {aka ‘struct _cr_PkgIterator’}                                                       
      535 |         cr_close(iter->other_f, NULL);                                                                                                                                                                                            
          |                      ^~                                                                                                                                                                                                           
    /home/vagrant/devel/community/createrepo_c/src/xml_parser_main_metadata_together.c:537:31: error: invalid use of incomplete typedef ‘cr_PkgIterator’ {aka ‘struct _cr_PkgIterator’}                                                       
      537 |         GError* tmp_err = iter->tmp_err;                                                                                                                                                                                          
          |                               ^~
    /home/vagrant/devel/community/createrepo_c/src/xml_parser_main_metadata_together.c:539:22: error: invalid use of incomplete typedef ‘cr_PkgIterator’ {aka ‘struct _cr_PkgIterator’}
      539 |         cr_close(iter->primary_f, &tmp_err);
          |                      ^~                                                                                    
    /home/vagrant/devel/community/createrepo_c/src/xml_parser_main_metadata_together.c:541:26: error: invalid use of incomplete typedef ‘cr_PkgIterator’ {aka ‘struct _cr_PkgIterator’}
      541 |             cr_close(iter->filelists_f, &tmp_err);
          |                          ^~
    /home/vagrant/devel/community/createrepo_c/src/xml_parser_main_metadata_together.c:544:26: error: invalid use of incomplete typedef ‘cr_PkgIterator’ {aka ‘struct _cr_PkgIterator’}
      544 |             cr_close(iter->other_f, &tmp_err);
          |                          ^~
```

Reply:
I just tried it and it seems to work for me.

Looking at the errors you mentioned make sure you are not `typedef`ing the `_cr_PkgIterator` in xml_parser_main_metadata_together.c.

It is working for me if I use the exact form I have in my comment: `struct _cr_PkgIterator {.. };` not `typedef struct {...} _cr_PkgIterator;`. So it cannot be just copied from xml_parser.h.

Reply:
>It is working for me if I use the exact form I have in my comment 

Yup, that's the issue, even without using `typedef` it seems like if you use the typedef order on the struct it doesn't work.